Add ID translation for Labels and Selectors (#19072)
kubernetes/website: content/id/docs/concepts/overview/working-with-objects/labels.md

---
reviewers:

> - mikedanese

Author: Ah, I missed it. Okay I'll remove it

> title: Label dan Selektor

Author: I've found "selector" in this glossarium from Pusat Bahasa http://bahasasastra.kemdikbud.go.id/glosarium/index.php?gloss_asing=selector&gloss_indonesia=&jenis=exact&Bidang=all&infocmd=Cari which is translated to "Selektor".

Besides, we also have other docs that using "Selektor" like this one https://kubernetes.io/id/docs/concepts/overview/working-with-objects/field-selectors/

> title: Label dan Selektor

Reviewer: Sorry for the really belated comments. Kinda heavy workloads lately. 
Thinking about it, as long as it's comprehensible to the readers, it's okay to "invent" a word.
We might also focus on getting pages done instead of being blocked by these types of problems.
Let's just use Selektor then!

{{% capture body %}}

## Motivasi

Label memungkinkan pengguna untuk memetakan struktur organisasi mereka ke dalam sistem objek yang tidak terikat erat, tanpa harus mewajibkan klien untuk menyimpan pemetaan tersebut.

_Service deployments_ dan _batch processing pipelines_ sering menjadi entitas yang multi-dimensional (contohnya multipel partisi atau _deployment_, multipel jalur rilis, multipel tingkatan, multipel _micro-services_ per tingkatan). Manajemen biasanya membutuhkan operasi yang lintas sektoral, yang menyebabkan putusnya enkapsulasi dari representasi hirarki, khususnya pada hirarki yang kaku yang ditentukan oleh infrastruktur, bukan malah oleh pengguna.

Contoh label:

   * `"release" : "stable"`, `"release" : "canary"`
   * `"environment" : "dev"`, `"environment" : "qa"`, `"environment" : "production"`
   * `"tier" : "frontend"`, `"tier" : "backend"`, `"tier" : "cache"`
   * `"partition" : "customerA"`, `"partition" : "customerB"`
   * `"track" : "daily"`, `"track" : "weekly"`

Ini hanya contoh label yang biasa digunakan; kamu bebas mengembangkan konvensimu sendiri. Perlu diingat bahwa label _Key_ harus unik untuk objek tersebut.

## Sintaksis dan set karakter

> _Label_ merupakan pasangan _key/value_. Label _Key_ yang valid memiliki dua segmen: sebuah prefiks yang opsional dan nama, dipisahkan oleh garis miring (`/`). Segmen nama wajib diisi dan harus memiliki 63 karakter atau kurang, dimulai dan diakhiri dengan karakter alfanumerik (`[a-z0-9A-Z]`) dengan tanda pisah (`-`), garis bawah (`_`), titik (`.`), dan alfanumerik di antaranya. Sedangkan prefiks opsional. Jika ditentukan, prefiks harus berupa subdomain DNS: rangkaian label DNS yang dipisahkan oleh titik (`.`), serta total tidak lebih dari 253 karakter, yang diikuti oleh garis miring (`/`).

Reviewer: ```suggestion
_Label_ merupakan pasangan _key/value_. _Key-key_ dari Label yang valid memiliki dua segmen: sebuah prefiks dan nama yang opsional, yang dipisahkan oleh garis miring (`/`). Segmen nama wajib diisi dan tidak boleh lebih dari 63, dimulai dan diakhiri dengan karakter alfanumerik (`[a-z0-9A-Z]`) dengan tanda pisah (`-`), garis bawah (`_`), titik (`.`), dan alfanumerik di antaranya. Sedangkan prefiks bersifat opsional. Jika ditentukan, prefiks harus berupa subdomain DNS: rangkaian label DNS yang dipisahkan oleh titik (`.`), dengan total tidak lebih dari 253 karakter, yang diikuti oleh garis miring (`/`).
```

> Jika prefiks dihilangkan, label _Key_ diasumsikan privat bagi pengguna. Komponen sistem otomatis (contoh `kube-scheduler`, `kube-controller-manager`, `kube-apiserver`, `kubectl`, atau otomasi pihak ketiga lainnya) yang akan menambah label ke objek pengguna akhir harus menspesifikasikan prefiks.

Reviewer: ```suggestion
Jika prefiks dihilangkan, _Key_ dari label diasumsikan privat bagi pengguna. Komponen sistem otomatis (contoh `kube-scheduler`, `kube-controller-manager`, `kube-apiserver`, `kubectl`, atau otomasi pihak ketiga lainnya) yang akan menambah label ke objek-objek milik pengguna akhir harus menentukan prefiks.
```

Prefiks `kubernetes.io/` dan `k8s.io/` dikhususkan untuk komponen inti Kubernetes.

> Nilai label yang valid harus terdiri dari 63 karakter atau kurang dan harus kosong atau diawali dan diakhiri dengan  karakter alfanumerik (`[a-z0-9A-Z]`) dengan tanda pisah (`-`), garis bawah (`_`), titik (`.`), dan alfanumerik di antaranya.

Reviewer: ```suggestion
Nilai label yang valid tidak boleh lebih dari 63 karakter dan harus kosong atau diawali dan diakhiri dengan karakter alfanumerik (`[a-z0-9A-Z]`) dengan tanda pisah (`-`), garis bawah (`_`), titik (`.`), dan alfanumerik di antaranya.
```

Contoh di bawah ini merupakan berkas konfigurasi untuk Pod yang memiliki dua label `environment: production` dan `app: nginx` :

```yaml

apiVersion: v1
kind: Pod
metadata:
  name: label-demo
  labels:
    environment: production
    app: nginx
spec:
  containers:
  - name: nginx
    image: nginx:1.7.9
    ports:
    - containerPort: 80

```

## Selektor label

Tidak seperti [nama dan UID](/id/docs/concepts/overview/working-with-objects/names/), label tidak memberikan keunikan. Secara umum, kami memperkirakan bahwa banyak objek yang akan memiliki label yang sama.

> Dengan sebuah _label selector_, klien/pengguna dapat mengidentifikasi suatu set objek. Selektor label merupakan pengelompokan primitif utama pada Kubernetes.

Reviewer: ```suggestion
Menggunakan sebuah _label selector_, klien/pengguna dapat mengidentifikasi suatu kumpulan objek. Selektor label merupakan alat/cara pengelompokan utama pada Kubernetes.
```

Saat ini API mendukung dua jenis selektor: _equality-based_ dan _set-based_.

> Sebuah selektor label dapat dibuat dari multipel kondisi yang dipisahkan oleh koma. Pada kasus multipel kondisi, semua kondisi harus dipenuhi sehingga separator koma dapat bertindak sebagai operator logika _AND_ (`&&`).

Reviewer: ```suggestion
Sebuah selektor label dapat dibuat dari kondisi berganda yang dipisahkan oleh koma. Pada kasus kondisi berganda, semua kondisi harus dipenuhi sehingga separator koma dapat bertindak sebagai operator logika _AND_ (`&&`).
```

> Semantik dari selektor yang kosong atau tidak dispesifikasikan tergantung dari konteks, dan tipe API yang menggunakan selektor harus mendokumentasikan kevalidan dan arti dari hal tersebut.

Reviewer: ```suggestion
Makna dari selektor yang kosong atau tidak diisi tergantung dari konteks, dan tipe API yang menggunakan selektor harus mendokumentasikan keabsahan dan arti dari selektor yang kosong tersebut.
```

{{< note >}}

> Untuk beberapa tipe API, seperti ReplicaSets, label selektor untuk dua _instance_ tidak boleh tumpang tindih dengan _namespace_, jika tidak maka _controller_ akan melihatnya sebagai instruksi yang menyebabkan konflik dan akan gagal menentukan berapa banyak replika yang harus tersedia.

Reviewer: ```suggestion
Untuk beberapa tipe API, seperti ReplicaSet, selektor label untuk dua objek tidak boleh tumpang tindih dengan Namespace, jika tidak maka _controller_ akan melihatnya sebagai instruksi yang menyebabkan konflik dan akan gagal menentukan berapa banyak replika yang seharusnya tersedia.
```

{{< /note >}}

{{< caution >}}

> Untuk kedua kondisi _equality-based_ dan _set-based_ tidak ada logika operator _OR_ (`||`). Pastikan struktur filter pernyataan kamu ikut disesuaikan.

Reviewer: ```suggestion
Untuk kedua kondisi _equality-based_ dan _set-based_ tidak ada logika operator _OR_ (`||`). Pastikan struktur pernyataan filter kamu ikut disesuaikan.
```

{{< /caution >}}

### Kondisi _Equality-based_

> Kondisi _Equality-based_ atau _inequality-based_ memungkinkan untuk melakukan filter dengan menggunakan label _key_ dan _value_. Objek yang cocok harus memenuhi semua batasan label yang dispesifikasikan, meskipun mereka dapat memiliki label tambahan lainnya.

Reviewer: ```suggestion
Kondisi _Equality-based_ atau _inequality-based_ memungkinkan untuk melakukan filter dengan menggunakan _key_ dan _value_ dari label. Objek yang cocok harus memenuhi semua batasan label yang telah ditentukan, meskipun mereka dapat memiliki label tambahan lainnya.
```

> Terdapat tiga jenis operator yang didukung yaitu `=`,`==`,`!=`. Dua operator pertama merepresentasikan _equality_ (dan hanya sinonim biasa), sementara operator terakhir merepresentasikan _inequality_. Contoh:

Reviewer: ```suggestion
Terdapat tiga jenis operator yang didukung yaitu `=`,`==`,`!=`. Dua operator pertama menyatakan kesamaan (keduanya hanyalah sinonim), sementara operator terakhir menyatakan ketidaksamaan. Contoh:
```

```
environment = production
tier != frontend
```

Kondisi pertama akan memilih semua sumber daya dengan _key_ `environment` dan nilai _key_ `production`.

> Kondisi berikutnya akan memilih semua sumber daya dengan _key_ `tier` dan nilai _key_ selain `frontend`, dan semua sumber daya yang tidak memiliki label _key_ `tier`.

Reviewer: ```suggestion
Kondisi berikutnya akan memilih semua sumber daya dengan _key_ `tier` dan nilai _key_ selain `frontend`, dan semua sumber daya yang tidak memiliki label dengan _key_ `tier`.
```

Kamu juga dapat memfilter sumber daya dalam `production` selain `frontend` dengan menggunakan operator koma: `environment=production,tier!=frontend`

> Salah satu skenario penggunaan label dengan kondisi _equality-based_ yaitu untuk kriteria pemilihan _node_ bagi Pods. Contoh, contoh Pod di bawah ini akan memilih _node_ dengan label "`accelerator=nvidia-tesla-p100`".

Reviewer: ```suggestion
Salah satu skenario penggunaan label dengan kondisi _equality-based_ yaitu untuk kriteria pemilihan Node untuk Pod-Pod. Sebagai contoh, Pod percontohan di bawah ini akan memilih Node dengan label "`accelerator=nvidia-tesla-p100`".
```

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: cuda-test
spec:
  containers:
    - name: cuda-test
      image: "k8s.gcr.io/cuda-vector-add:v0.1"
      resources:
        limits:
          nvidia.com/gpu: 1
  nodeSelector:
    accelerator: nvidia-tesla-p100
```

### Kondisi _Set-based_

> Kondisi label _Set-based_ memungkinkan melakukan filter _key_ terhadap suatu set nilai. Terdapat tiga jenis operator yang didukung yaitu: `in`,`notin` dan `exists` (hanya  untuk pengidentifikasian _key_). Contoh:

Reviewer: ```suggestion
Kondisi label _Set-based_ memungkinkan memfilter _key_ terhadap suatu kumpulan nilai. Terdapat tiga jenis operator yang didukung, yaitu: `in`,`notin`, dan `exists` (hanya  _key_-nya saja). Contoh:
```

```
environment in (production, qa)
tier notin (frontend, backend)
partition
!partition
```

Contoh pertama akan memilih semua sumber daya dengan _key_ `environment` dan nilai `production` atau `qa`.

> Contoh kedua akan memilih semua sumber daya dengan _key_ `tier` dan nilai selain `frontend` dan `backend`, serta semua sumber daya yang tidak memiliki label _key_ `tier`.

Reviewer: ```suggestion
Contoh kedua akan memilih semua sumber daya dengan _key_ `tier` dan nilai selain `frontend` dan `backend`, serta semua sumber daya yang tidak memiliki label dengan _key_ `tier`.
```

> Contoh ketiga akan memilih semua sumber daya yang memiliki label _key_ `partition`; tidak ada nilai yang diperiksa.

Reviewer: ```suggestion
Contoh ketiga akan memilih semua sumber daya yang memiliki _key_ dari label`partition`; nilainya tidak diperiksa.
```

> Sedangkan contoh contoh keempat akan memilih semua sumber daya yang tidak memiliki label `partition`; tidak ada nilai yang diperiksa.

Reviewer: ```suggestion
Sedangkan contoh keempat akan memilih semua sumber daya yang tidak memiliki label dengan _key_ `partition`; nilainya tidak diperiksa.
```

> Demikian pula untuk separator koma yang akan bertindak sebagai operator _AND_. Sehingga penyaringan sumber daya dengan _key_ `partition` (tidak peduli nilai dari _key_) dan `environment` yang tidak sama dengan `qa` dapat dicapai dengan `partition,environment notin (qa)`.

Reviewer: ```suggestion
Secara serupa, operator koma bertindak sebagai operator _AND_. Sehingga penyaringan sumber daya dengan _key_ `partition` (tidak peduli nilai dari _key_) dan `environment` yang tidak sama dengan `qa` dapat dicapai dengan `partition,environment notin (qa)`.
```

Selektor label _set-based_ merupakan bentuk umum persamaan karena `environment=production` sama dengan `environment in (production)`; demikian pula `!=` dan `notin`.

Kondisi _Set-based_ dapat digabungkan dengan kondisi _equality-based_. Contoh: `partition in (customerA, customerB),environment!=qa`.

## API

### Penyaringan LIST dan WATCH

> Operasi LIST dan WATCH dapat digunakan untuk menspesifikasikan selektor label untuk memfilter suatu set objek yang didapat dengan menggunakan parameter kueri. Kedua kondisi diperbolehkan (yang ditampilkan di sini seperti saat tampil pada kueri string di URL):

Reviewer: ```suggestion
Operasi LIST dan WATCH dapat menentukan selektor label untuk memfilter suatu kumpulan objek yang didapat dengan menggunakan parameter kueri. Kedua jenis kondisi diperbolehkan (ditampilkan sebagai berikut, sama seperti saat tampil pada string kueri di URL):
```

  * Kondisi _equality-based_: `?labelSelector=environment%3Dproduction,tier%3Dfrontend`
  * Kondisi _set-based_: `?labelSelector=environment+in+%28production%2Cqa%29%2Ctier+in+%28frontend%29`

> Kedua jenis selektor label dapat digunakan untuk menampilkan dan mengamati (_watch_) sumber daya melalui klien REST. Contoh, untuk menargetkan `apiserver` dengan `kubectl` dan menggunakan _equality-based_ kamu dapat menuliskan:

Reviewer: ```suggestion
Kedua jenis selektor label dapat digunakan untuk menampilkan (_list_) dan mengamati (_watch_) sumber daya melalui klien REST. Contohnya, menargetkan `apiserver` dengan `kubectl` dan menggunakan _equality-based_ kamu dapat menuliskan:
```

```shell
kubectl get pods -l environment=production,tier=frontend
```

> atau menggunakan kondisi or using _set-based_:

Reviewer: ```suggestion
atau menggunakan kondisi _set-based_:
```

```shell
kubectl get pods -l 'environment in (production),tier in (frontend)'
```

> Seperti yang telah disebutkan sebelumnya, kondisi _set-based_ lebih ekspresif.  Sebagai contoh, mereka dapat digunakan untuk mengimplementasikan operator _OR_ pada nilai:

Reviewer: ```suggestion
Seperti yang telah disebutkan sebelumnya, kondisi _set-based_ lebih ekspresif.  Sebagai contoh, mereka dapat digunakan untuk mengimplementasi operator _OR_ pada nilai:
```

```shell
kubectl get pods -l 'environment in (production, qa)'
```

atau membatasi pencocokan negatif dengan operator _exists_:

```shell
kubectl get pods -l 'environment,environment notin (frontend)'
```

### Mengatur referensi pada objek API

> Pada beberapa objek Kubernetes, seperti [`services`](/docs/user-guide/services) dan [`replicationcontrollers`](/id/docs/concepts/workloads/controllers/replicationcontroller/), juga menggunakan selektor label untuk menspesifikasikan set dari sumber daya lain, seperti [pods](/id/docs/concepts/workloads/pods/pod).

Reviewer: ```suggestion
Pada beberapa objek Kubernetes, seperti [`Service`](/docs/user-guide/services) dan [`ReplicationController`](/id/docs/concepts/workloads/controllers/replicationcontroller/), juga menggunakan selektor label untuk menentukan kumpulan dari sumber daya lain, seperti [Pod](/id/docs/concepts/workloads/pods/pod).
```

#### Service dan ReplicationController

> Set Pod yang ditargetkan oleh `service` didefinisikan oleh selektor label. Demikian pula populasi pod yang ditangani oleh `replicationcontroller` juga didefinisikan dengan selektor label.

Reviewer: ```suggestion
Kumpulan Pod yang ditargetkan oleh sebuah `service` ditentukan dengan selektor label. Demikian pula kumpulan Pod yang harus ditangani oleh `replicationcontroller` juga ditentukan dengan selektor label.
```

> Selektor label untuk kedua objek didefinisikan dalam berkas `json` atau `yaml` menggunakan _maps_, dan hanya mendukung kondisi _equality-based_:

Reviewer: ```suggestion
Selektor label untuk kedua objek tersebut ditentukan dalam berkas `json` atau `yaml` menggunakan _maps_, dan hanya mendukung kondisi _equality-based_:
```

```json
"selector": {
    "component" : "redis",
}
```
atau

```yaml
selector:
    component: redis
```

> selektor ini (baik dalam bentuk format `json` atau `yaml`) sama dengan `component=redis` atau `component in (redis)`.

Reviewer: ```suggestion
selektor ini (baik dalam bentuk `json` atau `yaml`) sama dengan `component=redis` atau `component in (redis)`.
```

#### Sumber daya yang mendukung kondisi set-based

Sumber daya yang lebih baru, seperti [`Job`](/id/docs/concepts/workloads/controllers/jobs-run-to-completion/), [`Deployment`](/id/docs/concepts/workloads/controllers/deployment/), [`ReplicaSet`](/id/docs/concepts/workloads/controllers/replicaset/), dan [`DaemonSet`](/id/docs/concepts/workloads/controllers/daemonset/), juga mendukung kondisi _set-based_.

```yaml
selector:
  matchLabels:
    component: redis
  matchExpressions:
    - {key: tier, operator: In, values: [cache]}
    - {key: environment, operator: NotIn, values: [dev]}
```

> `matchLabels` merupakan pemetaan dari pasangan `{key,value}`. Sebuah `{key,value}` pada pemetaan `matchLabels` adalah sama dengan elemen dari `matchExpressions`, yang nilai `key` nya adalah "key", dengan `operator` yaitu "In", dengan nilai `values` array hanya mengandung "value". `matchExpressions` merupakan kebutuhan untuk selektor pod. Operator yang valid termasuk In, NotIn, Exists, dan DoesNotExist. Nilai ini harus diatur tidak boleh kosong pada kasus In dan NotIn. Semua kondisi, baik dari  `matchLabels` dan `matchExpressions` di-ANDed bersama-sama -- mereka harus memenuhi semua kondisi agar cocok.

Reviewer: ```suggestion
`matchLabels` merupakan pemetaan dari pasangan `{key,value}`. Sebuah `{key,value}` pada pemetaan `matchLabels` adalah sama dengan elemen dari `matchExpressions`, yang nilai `key` nya adalah "key", dengan `operator` "In", dan _array_ `values` hanya berisi "value". `matchExpressions` merupakan daftar kondisi untuk selektor Pod. Operator yang valid termasuk In, NotIn, Exists, dan DoesNotExist. Kumpulan nilai ini tidak boleh kosong pada kasus In dan NotIn. Semua kondisi, baik dari  `matchLabels` dan `matchExpressions` di-AND secara sekaligus -- mereka harus memenuhi semua kondisi agar cocok.
```

> #### Memilih set dari node

Reviewer: ```suggestion
#### Memilih kumpulan Node
```

> Salah satu contoh penggunaan pemilihan dengan menggunakan label yaitu untuk membatasi suatu set _node_ tertentu yang dapat digunakan oleh pod.

Reviewer: ```suggestion
Salah satu contoh penggunaan pemilihan dengan menggunakan label yaitu untuk membatasi suatu kumpulan Node tertentu yang dapat digunakan oleh Pod.
```

> Lihat dokumentasi pada [pemilihan _node_](/id/docs/concepts/configuration/assign-pod-node/) untuk informasi lebih lanjut.

Reviewer: ```suggestion
Lihat dokumentasi pada [pemilihan Node](/id/docs/concepts/configuration/assign-pod-node/) untuk informasi lebih lanjut.
```
